### ISSUE TITLE
Upgrade with AMD64 WPK package macOS modifies ARM64 agent and generates a failure in its operation.

### ISSUE THREAD

**Issue by pro-akim** [wazuh/wazuh#26455]
| Version | Revision | Production/Development | Component | Install type |
|:--:|:--:|:--:|:--:|:--:|
|  v4.10.0  |  1  | Development   | macOS Agent   | Download/Repository/Sources/Assistant | 

## Description 

During the tests performed at https://github.com/wazuh/wazuh/issues/26434, the test macOS Ventura ARM WPK upgrade failed. I
The test consists of forcing a macOS ARM 4.9.1 Agent using WPK for AMD to upgrade to 4.10.0.

### Environment

- AMI v4.10.0-alpha2 (AWS)
- macOS ARM64 Ventura 13 Agent (Deployed via Allocation module) (Mac Mini)

### Steps to reproduce

- Install and register a 4.9.1 Agent to a 4.10.0 Manager.
- The Agent must be macOS ARM64 Ventura 13.
- After that, download the WPK package for macOS AMD64 and run the upgrade.

### Current result

The process is blocked for an unlimited time.
When performing a manual stop, it is observed that the Agent is disconnected.
The Agent, on the other hand, has done the update (which it shouldn't) and it doesn't work.
It has errors in the log

```console
[root@wazuh-server wazuh-user]# /var/ossec/bin/agent_upgrade -a 014 -f /home/wazuh-user/wazuh_agent_v4.10.0_macos_intel64.pkg.wpk 

Upgrading...
^Z
[2]+  Stopped                 /var/ossec/bin/agent_upgrade -a 014 -f /home/wazuh-user/wazuh_agent_v4.10.0_macos_intel64.pkg.wpk
[root@wazuh-server wazuh-user]# /var/ossec/bin/agent_control -l

Wazuh agent_control. List of available agents:
   ID: 000, Name: wazuh-server (server), IP: 127.0.0.1, Active/Local
   ID: 014, Name: macos-13, IP: any, Disconnected

List of agentless devices:
```

```console
sh-3.2# /Library/Ossec/bin/wazuh-control info
WAZUH_VERSION="v4.10.0"
WAZUH_REVISION="41003"
WAZUH_TYPE="agent"

sh-3.2# /Library/Ossec/bin/wazuh-control restart
/Library/Ossec/bin/wazuh-control: line 120: /Library/Ossec/bin/wazuh-execd: Bad CPU type in executable
wazuh-execd: Configuration error. Exiting

```

- Check error in `/Library/Ossec/logs/upgrade.log`

```console
sh-3.2# cat /Library/Ossec/logs/upgrade.log
2024/10/21 06:36:37 - Upgrade started.
2024/10/21 06:36:37 - Checking execution path.
installer: Package name is wazuh-agent-4.10.0-1.intel64
installer: Upgrading at base path /
installer: The upgrade failed. (The Installer encountered an error that caused the installation to fail. Contact the software manufacturer for assistance. An error occurred while running scripts from the package “wazuh-agent-4.10.0-1.intel64.pkg”.)
2024/10/21 06:36:40 - Installation result = 1
2024/10/21 06:36:40 - Checking for Wazuh Agent control script.
2024/10/21 06:36:40 - Restarting Wazuh Agent.
./bin/wazuh-control: line 120: /Library/Ossec/bin/wazuh-execd: Bad CPU type in executable
wazuh-execd: Configuration error. Exiting
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:41 - Waiting connection... Remaining attempts: 30.
2024/10/21 06:36:42 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:42 - Waiting connection... Remaining attempts: 29.
2024/10/21 06:36:43 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:43 - Waiting connection... Remaining attempts: 28.
2024/10/21 06:36:44 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:44 - Waiting connection... Remaining attempts: 27.
2024/10/21 06:36:45 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:45 - Waiting connection... Remaining attempts: 26.
2024/10/21 06:36:46 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:46 - Waiting connection... Remaining attempts: 25.
2024/10/21 06:36:47 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:47 - Waiting connection... Remaining attempts: 24.
2024/10/21 06:36:48 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:48 - Waiting connection... Remaining attempts: 23.
2024/10/21 06:36:49 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:49 - Waiting connection... Remaining attempts: 22.
2024/10/21 06:36:50 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:50 - Waiting connection... Remaining attempts: 21.
2024/10/21 06:36:51 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:51 - Waiting connection... Remaining attempts: 20.
2024/10/21 06:36:52 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:52 - Waiting connection... Remaining attempts: 19.
2024/10/21 06:36:53 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:53 - Waiting connection... Remaining attempts: 18.
2024/10/21 06:36:54 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:54 - Waiting connection... Remaining attempts: 17.
2024/10/21 06:36:55 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:55 - Waiting connection... Remaining attempts: 16.
2024/10/21 06:36:56 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:56 - Waiting connection... Remaining attempts: 15.
2024/10/21 06:36:57 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:57 - Waiting connection... Remaining attempts: 14.
2024/10/21 06:36:58 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:36:59 - Waiting connection... Remaining attempts: 13.
2024/10/21 06:37:00 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:37:00 - Waiting connection... Remaining attempts: 12.
2024/10/21 06:37:01 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:37:01 - Waiting connection... Remaining attempts: 11.
2024/10/21 06:37:02 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:37:02 - Waiting connection... Remaining attempts: 10.
2024/10/21 06:37:03 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:37:03 - Waiting connection... Remaining attempts: 9.
2024/10/21 06:37:04 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:37:04 - Waiting connection... Remaining attempts: 8.
2024/10/21 06:37:05 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:37:05 - Waiting connection... Remaining attempts: 7.
2024/10/21 06:37:06 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:37:06 - Waiting connection... Remaining attempts: 6.
2024/10/21 06:37:07 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:37:07 - Waiting connection... Remaining attempts: 5.
2024/10/21 06:37:08 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:37:08 - Waiting connection... Remaining attempts: 4.
2024/10/21 06:37:09 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:37:09 - Waiting connection... Remaining attempts: 3.
2024/10/21 06:37:10 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:37:10 - Waiting connection... Remaining attempts: 2.
2024/10/21 06:37:11 - Status = pending. 
./var/upgrade/pkg_installer.sh: line 124: ./var/run/wazuh-agentd.state: No such file or directory
2024/10/21 06:37:11 - Waiting connection... Remaining attempts: 1.
2024/10/21 06:37:12 - Status = pending. 
2024/10/21 06:37:12 - Upgrade failed...
```

```console
sh-3.2# ps aux | grep wazuh 
root             32147   0,0  0,0 408626848   1232 s000  R+    6:51AM   0:00.01 grep wazuh
```

### Expected result
The agent should not update and a message should appear stating that it cannot.

**Comment by vikman90** [comment]
Hi @pro-akim,

Please note that as of version 4.10.0, the Wazuh agent **no longer officially supports macOS Ventura**. Our current supported versions are macOS Sonoma and Sequoia.

I have followed the steps you provided on macOS 15 Sonoma (ARM), and the upgrade completes without any issues:

### Agent-side

```shell
echo "WAZUH_MANAGER='192.168.1.3'" > /tmp/wazuh_envs && installer -pkg wazuh-agent-4.9.1-1.arm64.pkg -target /
/Library/Ossec/bin/wazuh-control start
```

### Manager-side

```shell
/var/ossec/bin/agent_control -i 001
```
> Wazuh agent_control. Agent information:
   Agent ID:   001
   Agent Name: Vikmans-MacBook-Pro.local
   IP address: any
   Status:     Active
>
>   Operating system:    Darwin |Vikmans-MacBook-Pro.local |24.0.0 |Darwin Kernel Version 24.0.0: Tue Sep 24 23:37:25 PDT 2024; root:xnu-11215.1.12~1/RELEASE_ARM64_T6030 |arm64
   Client version:      Wazuh v4.9.1
   Configuration hash:  ab73af41699f13fdd81903b5f23d8d00
   Shared file hash:    732f0957789452a4dbbb7e9a84480650
   Last keep alive:     1729581011
>
>   Syscheck last started at:  Tue Oct 22 09:07:41 2024
   Syscheck last ended at:    Tue Oct 22 09:07:42 2024

```shell
/var/ossec/bin/agent_upgrade -a 001 -f /root/wazuh_agent_v4.10.0_macos_intel64.pkg.wpk
```

> Upgrading...
>
> Upgraded agents:
        Agent 001 upgraded: Wazuh v4.9.1 -> Wazuh v4.10.0

## Possible causes

It is likely that this problem is related to changes introduced in version 4.10.0, including:
- https://github.com/wazuh/wazuh/pull/25091
- https://github.com/wazuh/wazuh/pull/23760

## Conclusion

Since the issue does not occur in macOS Sonoma or Sequoia, and we will be conducting additional manual tests on our side (#26458), we suggest continuing the rest of your testing on those versions and **discarding Ventura for version 4.10.0**. However, if you encounter the same problem on Sonoma or Sequoia, please **reopen this issue and provide access to the environment so we can investigate further**.

Best regards.

**Comment by vikman90** [comment]
With the help of @pro-akim, we have identified the root cause of the issue: Rosetta is not installed on this macOS system.

This problem is not specifically related to WPK and is also reproducible with PKG packages. Furthermore, it does not depend on the agent version (4.10.0) or macOS version (Ventura). The issue only arises when the package is Intel64, the system is ARM64, and Rosetta is not installed.

One way to check if Rosetta is installed is by running:

```
arch -x86-64 zsh -c ''
```

If Rosetta is installed, the command will exit with code 0. If Rosetta is not installed, it will result in an error such as:
> arch: posix_spawnp: zsh: Bad CPU type in executable

## Fix proposal

A possible fix would be to add a pre-installation check to verify that Rosetta is installed when the system is ARM64 and the package is Intel64.

**Comment by vikman90** [comment]
## Triage update

We've tried to install both 4.9.1 and 4.10.0. While the former prints that the package requires Rosetta and aborts, the latter continues installing.

### 4.9.1

```shell
sudo installer -pkg ./wazuh-agent-4.9.1-1.intel64.pkg -target /
```

> installer: This package requires Rosetta 2 to be installed.
Please install Rosetta 2 and then try again.
`sudo softwareupdate --install-rosetta`

### 4.10.0 Alpha 2

```shell
sudo installer -pkg ./wazuh-agent-4.10.0-1.intel64.pkg -target /
```

> installer: Package name is wazuh-agent-4.10.0-1.intel64
installer: Installing at base path /
installer: The install was successful.

```shell
sudo /Library/Ossec/bin/wazuh-agentd -t
```

> sudo: unable to execute /Library/Ossec/bin/wazuh-agentd: Bad CPU type in executable

## Conclusion

This is a regression, and we should fix this in 4.10.0-alpha3. We need to declare the package architecture in the SPECs.

**Comment by vikman90** [comment]
## Work report

- I added the architecture check to the pre-install script.
- I added a constant that will populate the package generator.
- PR https://github.com/wazuh/wazuh/pull/26480
- There is a "x86-64" typo for "x86_64", fixed but pending review.
